File: Music-IBook-API/Music-IBook-API/Services/TeacherService.cs
using Microsoft.EntityFrameworkCore;
using Music_IBook_API.DTOs;
using Music_IBook_API.Models;

namespace Music_IBook_API.Services;

public class TeacherService : ITeacherService
{
    private readonly MusicIBookDbContext db;

    public TeacherService(MusicIBookDbContext db)
    {
        this.db = db;
    }

    public async Task<TeacherDashboardDto> GetDashboardAsync(long teacherId)
    {
        var totalLessons = await db.Lessons.CountAsync(x => x.TeacherId == teacherId);
        var totalStudents = await db.Users.CountAsync(x => x.Role == "Student");
        var totalPracticeSessions = await db.PracticeSessions.CountAsync(x => !x.IsExam);
        var totalExamSessions = await db.PracticeSessions.CountAsync(x => x.IsExam);
        
        var avgScore = await db.PracticeSessions
            .Where(x => x.IsExam)
            .AverageAsync(x => (double?)x.Score) ?? 0;

        return new TeacherDashboardDto
        {
            TotalLessons = totalLessons,
            TotalStudents = totalStudents,
            TotalPracticeSessions = totalPracticeSessions,
            TotalExamSessions = totalExamSessions,
            AverageSystemScore = Math.Round(avgScore, 2)
        };
    }

    public async Task<List<StudentProgressOverviewDto>> GetStudentsProgressAsync()
    {
        var students = await db.Users
            .Where(x => x.Role == "Student")
            .Select(s => new StudentProgressOverviewDto
            {
                StudentId = s.Id,
                StudentName = s.FullName,
                Email = s.Email,
                LessonCount = db.PracticeSessions.Where(ps => ps.StudentId == s.Id && !ps.IsExam).Select(ps => ps.LessonId).Distinct().Count(),
                ExamCount = db.PracticeSessions.Where(ps => ps.StudentId == s.Id && ps.IsExam).Count(),
                AverageScore = db.PracticeSessions.Where(ps => ps.StudentId == s.Id && ps.IsExam).Average(ps => (double?)ps.Score) ?? 0,
                LastActivityAt = db.PracticeSessions
                    .Where(ps => ps.StudentId == s.Id)
                    .Max(ps => (DateTime?)(ps.FinishedAtUtc ?? ps.StartedAtUtc))
            })
            .ToListAsync();

        foreach (var student in students)
            student.LastActivityAt = AsUtc(student.LastActivityAt);
        return students;
    }

    public async Task<StudentDetailProgressDto> GetStudentProgressDetailAsync(long studentId)
    {
        var student = await db.Users.FindAsync(studentId);
        if (student == null) throw new Exception("Không tìm thấy học sinh");

        var lessons = await db.Lessons.ToListAsync();
        var sessions = await db.PracticeSessions
            .Where(x => x.StudentId == studentId)
            .ToListAsync();

        var lessonProgresses = new List<LessonProgressDetailDto>();

        foreach (var lesson in lessons)
        {
            var lessonSessions = sessions.Where(s => s.LessonId == lesson.Id).ToList();
            var practiceSessions = lessonSessions.Where(s => !s.IsExam).ToList();
            var examSessions = lessonSessions.Where(s => s.IsExam).ToList();

            if (lessonSessions.Any())
            {
                lessonProgresses.Add(new LessonProgressDetailDto
                {
                    LessonId = lesson.Id,
                    LessonName = lesson.Title,
                    LastPracticeAt = practiceSessions.Any()
                        ? AsUtc(practiceSessions.Max(x => x.FinishedAtUtc ?? x.StartedAtUtc))
                        : null,
                    LastExamAt = examSessions.Any()
                        ? AsUtc(examSessions.Max(x => x.FinishedAtUtc ?? x.StartedAtUtc))
                        : null,
                    HighestScore = examSessions.Any() ? examSessions.Max(x => x.Score) : 0,
                    AverageScore = examSessions.Any() ? Math.Round(examSessions.Average(x => x.Score), 2) : 0,
                    Accuracy = examSessions.Any() ? Math.Round(examSessions.Average(x => x.Accuracy), 2) : 0
                });
            }
        }

        return new StudentDetailProgressDto
        {
            StudentId = student.Id,
            StudentName = student.FullName,
            Lessons = lessonProgresses.OrderByDescending(x => x.LastExamAt ?? x.LastPracticeAt).ToList()
        };
    }

    public async Task<LessonAnalyticsDto> GetLessonAnalyticsAsync(long studentId, long lessonId)
    {
        var studentExists = await db.Users.AnyAsync(x => x.Id == studentId && x.Role == "Student");
        var lessonExists = await db.Lessons.AnyAsync(x => x.Id == lessonId);
        if (!studentExists || !lessonExists)
            throw new Exception("Không tìm thấy học sinh hoặc bài học");

        var sessions = await db.PracticeSessions
            .Where(x => x.StudentId == studentId && x.LessonId == lessonId)
            .OrderByDescending(x => x.StartedAtUtc)
            .Take(30)
            .Select(x => new PracticeSessionTrendDto
            {
                SessionId = x.Id,
                IsExam = x.IsExam,
                Score = x.Score,
                Accuracy = x.Accuracy,
                CorrectCount = x.CorrectCount,
                WrongCount = x.WrongCount,
                StartedAtUtc = x.StartedAtUtc
            })
            .ToListAsync();

        sessions.Reverse();
        foreach (var session in sessions)
            session.StartedAtUtc = AsUtc(session.StartedAtUtc)!.Value;

        var errorNotes = await db.StudentNoteAttempts
            .Where(x => x.PracticeSession.StudentId == studentId &&
                        x.PracticeSession.LessonId == lessonId &&
                        !x.IsCorrect)
            .GroupBy(x => x.ExpectedNote)
            .Select(x => new NoteErrorDto
            {
                Note = x.Key,
                ErrorCount = x.Count(),
                WrongPitchCount = x.Count(a => !a.IsCorrectPitch),
                TimingErrorCount = x.Count(a => !a.IsCorrectTiming)
            })
            .OrderByDescending(x => x.ErrorCount)
            .ThenBy(x => x.Note)
            .Take(12)
            .ToListAsync();

        var assignments = await db.StudentAssignments
            .Where(x => x.StudentId == studentId && x.LessonId == lessonId)
            .OrderByDescending(x => x.CreatedAtUtc)
            .Select(x => new AssignmentDto
            {
                Id = x.Id,
                StudentId = x.StudentId,
                LessonId = x.LessonId,
                LessonSectionId = x.LessonSectionId,
                LessonExerciseId = x.LessonExerciseId,
                SectionTitle = x.Section == null ? null : x.Section.Title,
                ExerciseTitle = x.Exercise == null ? null : x.Exercise.Title,
                Message = x.Message,
                DueAtUtc = x.DueAtUtc,
                CreatedAtUtc = x.CreatedAtUtc,
                IsCompleted = x.IsCompleted
            })
            .ToListAsync();

        foreach (var assignment in assignments)
        {
            assignment.DueAtUtc = AsUtc(assignment.DueAtUtc);
            assignment.CreatedAtUtc = AsUtc(assignment.CreatedAtUtc)!.Value;
        }

        return new LessonAnalyticsDto
        {
            StudentId = studentId,
            LessonId = lessonId,
            Sessions = sessions,
            ErrorNotes = errorNotes,
            Assignments = assignments
        };
    }

    public async Task<AssignmentDto> CreateStudentAssignmentAsync(CreateStudentAssignmentRequest request)
    {
        var studentExists = await db.Users.AnyAsync(x => x.Id == request.StudentId && x.Role == "Student");
        var lessonExists = await db.Lessons.AnyAsync(x => x.Id == request.LessonId);
        if (!studentExists || !lessonExists)
            throw new Exception("Không tìm thấy học sinh hoặc bài học");

        if (request.LessonSectionId.HasValue && !await db.LessonSections.AnyAsync(x =>
                x.Id == request.LessonSectionId.Value && x.LessonId == request.LessonId))
            throw new ArgumentException("Đoạn luyện tập không thuộc bài học đã chọn.");

        if (request.LessonExerciseId.HasValue && !await db.LessonExercises.AnyAsync(x =>
                x.Id == request.LessonExerciseId.Value && x.LessonId == request.LessonId))
            throw new ArgumentException("Bài tập không thuộc bài học đã chọn.");

        var assignment = new StudentAssignment
        {
            StudentId = request.StudentId,
            LessonId = request.LessonId,
            LessonSectionId = request.LessonSectionId,
            LessonExerciseId = request.LessonExerciseId,
            Message = request.Message.Trim(),
            DueAtUtc = request.DueAtUtc,
            CreatedAtUtc = DateTime.UtcNow
        };
        db.StudentAssignments.Add(assignment);
        await db.SaveChangesAsync();

        return new AssignmentDto
        {
            Id = assignment.Id,
            StudentId = assignment.StudentId,
            LessonId = assignment.LessonId,
            LessonSectionId = assignment.LessonSectionId,
            LessonExerciseId = assignment.LessonExerciseId,
            Message = assignment.Message,
            DueAtUtc = assignment.DueAtUtc,
            CreatedAtUtc = assignment.CreatedAtUtc,
            IsCompleted = assignment.IsCompleted
        };
    }

    private static DateTime? AsUtc(DateTime? value) => value.HasValue
        ? DateTime.SpecifyKind(value.Value, DateTimeKind.Utc)
        : null;
}

File: music_ibook_flutter/lib/services/teacher_service.dart
import '../core/api_client.dart';
import '../utils/vietnam_time.dart';

class TeacherDashboardData {
  final int totalLessons;
  final int totalStudents;
  final int totalPracticeSessions;
  final int totalExamSessions;
  final double averageSystemScore;

  TeacherDashboardData({
    required this.totalLessons,
    required this.totalStudents,
    required this.totalPracticeSessions,
    required this.totalExamSessions,
    required this.averageSystemScore,
  });

  factory TeacherDashboardData.fromJson(Map<String, dynamic> json) {
    return TeacherDashboardData(
      totalLessons: json['totalLessons'] ?? 0,
      totalStudents: json['totalStudents'] ?? 0,
      totalPracticeSessions: json['totalPracticeSessions'] ?? 0,
      totalExamSessions: json['totalExamSessions'] ?? 0,
      averageSystemScore: (json['averageSystemScore'] ?? 0).toDouble(),
    );
  }
}

class StudentProgressOverview {
  final int studentId;
  final String studentName;
  final String email;
  final int lessonCount;
  final int examCount;
  final double averageScore;
  final DateTime? lastActivityAt;

  StudentProgressOverview({
    required this.studentId,
    required this.studentName,
    required this.email,
    required this.lessonCount,
    required this.examCount,
    required this.averageScore,
    this.lastActivityAt,
  });

  factory StudentProgressOverview.fromJson(Map<String, dynamic> json) {
    return StudentProgressOverview(
      studentId: json['studentId'] ?? 0,
      studentName: json['studentName'] ?? '',
      email: json['email'] ?? '',
      lessonCount: json['lessonCount'] ?? 0,
      examCount: json['examCount'] ?? 0,
      averageScore: (json['averageScore'] ?? 0).toDouble(),
      lastActivityAt: VietnamTime.parseUtc(json['lastActivityAt']),
    );
  }
}

class LessonProgressDetail {
  final int lessonId;
  final String lessonName;
  final DateTime? lastPracticeAt;
  final DateTime? lastExamAt;
  final int highestScore;
  final double averageScore;
  final double accuracy;

  LessonProgressDetail({
    required this.lessonId,
    required this.lessonName,
    this.lastPracticeAt,
    this.lastExamAt,
    required this.highestScore,
    required this.averageScore,
    required this.accuracy,
  });

  factory LessonProgressDetail.fromJson(Map<String, dynamic> json) {
    return LessonProgressDetail(
      lessonId: json['lessonId'] ?? 0,
      lessonName: json['lessonName'] ?? '',
      lastPracticeAt: VietnamTime.parseUtc(json['lastPracticeAt']),
      lastExamAt: VietnamTime.parseUtc(json['lastExamAt']),
      highestScore: json['highestScore'] ?? 0,
      averageScore: (json['averageScore'] ?? 0).toDouble(),
      accuracy: (json['accuracy'] ?? 0).toDouble(),
    );
  }
}

class StudentDetailProgress {
  final int studentId;
  final String studentName;
  final List<LessonProgressDetail> lessons;

  StudentDetailProgress({
    required this.studentId,
    required this.studentName,
    required this.lessons,
  });

  factory StudentDetailProgress.fromJson(Map<String, dynamic> json) {
    var list = json['lessons'] as List? ?? [];
    return StudentDetailProgress(
      studentId: json['studentId'] ?? 0,
      studentName: json['studentName'] ?? '',
      lessons: list.map((e) => LessonProgressDetail.fromJson(e)).toList(),
    );
  }
}

class PracticeSessionTrend {
  final int sessionId;
  final bool isExam;
  final int score;
  final double accuracy;
  final int correctCount;
  final int wrongCount;
  final DateTime startedAt;

  PracticeSessionTrend({
    required this.sessionId,
    required this.isExam,
    required this.score,
    required this.accuracy,
    required this.correctCount,
    required this.wrongCount,
    required this.startedAt,
  });

  factory PracticeSessionTrend.fromJson(Map<String, dynamic> json) =>
      PracticeSessionTrend(
        sessionId: json['sessionId'] ?? 0,
        isExam: json['isExam'] ?? false,
        score: json['score'] ?? 0,
        accuracy: (json['accuracy'] ?? 0).toDouble(),
        correctCount: json['correctCount'] ?? 0,
        wrongCount: json['wrongCount'] ?? 0,
        startedAt:
            VietnamTime.parseUtc(json['startedAtUtc']) ??
            DateTime.fromMillisecondsSinceEpoch(0),
      );
}

class NoteErrorSummary {
  final String note;
  final int errorCount;
  final int wrongPitchCount;
  final int timingErrorCount;

  NoteErrorSummary({
    required this.note,
    required this.errorCount,
    required this.wrongPitchCount,
    required this.timingErrorCount,
  });

  factory NoteErrorSummary.fromJson(Map<String, dynamic> json) =>
      NoteErrorSummary(
        note: json['note'] ?? '',
        errorCount: json['errorCount'] ?? 0,
        wrongPitchCount: json['wrongPitchCount'] ?? 0,
        timingErrorCount: json['timingErrorCount'] ?? 0,
      );
}

class StudentAssignment {
  final int id;
  final int? lessonSectionId;
  final int? lessonExerciseId;
  final String? sectionTitle;
  final String? exerciseTitle;
  final String message;
  final DateTime? dueAt;
  final DateTime createdAt;
  final bool isCompleted;

  StudentAssignment({
    required this.id,
    this.lessonSectionId,
    this.lessonExerciseId,
    this.sectionTitle,
    this.exerciseTitle,
    required this.message,
    this.dueAt,
    required this.createdAt,
    required this.isCompleted,
  });

  factory StudentAssignment.fromJson(Map<String, dynamic> json) =>
      StudentAssignment(
        id: json['id'] ?? 0,
        lessonSectionId: json['lessonSectionId'],
        lessonExerciseId: json['lessonExerciseId'],
        sectionTitle: json['sectionTitle'],
        exerciseTitle: json['exerciseTitle'],
        message: json['message'] ?? '',
        dueAt: VietnamTime.parseUtc(json['dueAtUtc']),
        createdAt:
            VietnamTime.parseUtc(json['createdAtUtc']) ??
            DateTime.fromMillisecondsSinceEpoch(0),
        isCompleted: json['isCompleted'] ?? false,
      );
}

class LessonAnalytics {
  final List<PracticeSessionTrend> sessions;
  final List<NoteErrorSummary> errorNotes;
  final List<StudentAssignment> assignments;

  LessonAnalytics({
    required this.sessions,
    required this.errorNotes,
    required this.assignments,
  });

  factory LessonAnalytics.fromJson(Map<String, dynamic> json) =>
      LessonAnalytics(
        sessions: (json['sessions'] as List? ?? [])
            .map((item) => PracticeSessionTrend.fromJson(item))
            .toList(),
        errorNotes: (json['errorNotes'] as List? ?? [])
            .map((item) => NoteErrorSummary.fromJson(item))
            .toList(),
        assignments: (json['assignments'] as List? ?? [])
            .map((item) => StudentAssignment.fromJson(item))
            .toList(),
      );
}

class TeacherService {
  final _dio = ApiClient.instance.dio;

  Future<TeacherDashboardData> getDashboard() async {
    final res = await _dio.get('/api/teacher/dashboard');
    return TeacherDashboardData.fromJson(res.data);
  }

  Future<List<StudentProgressOverview>> getStudentsProgress() async {
    final res = await _dio.get('/api/teacher/students-progress');
    final list = res.data as List;
    return list.map((e) => StudentProgressOverview.fromJson(e)).toList();
  }

  Future<StudentDetailProgress> getStudentProgressDetail(int studentId) async {
    final res = await _dio.get('/api/teacher/students/$studentId/progress');
    return StudentDetailProgress.fromJson(res.data);
  }

  Future<LessonAnalytics> getLessonAnalytics(
    int studentId,
    int lessonId,
  ) async {
    final res = await _dio.get(
      '/api/teacher/students/$studentId/lessons/$lessonId/analytics',
    );
    return LessonAnalytics.fromJson(res.data);
  }

  Future<void> createAssignment({
    required int studentId,
    required int lessonId,
    int? lessonSectionId,
    int? lessonExerciseId,
    required String message,
    DateTime? dueAt,
  }) async {
    await _dio.post(
      '/api/teacher/assignments',
      data: {
        'studentId': studentId,
        'lessonId': lessonId,
        'lessonSectionId': lessonSectionId,
        'lessonExerciseId': lessonExerciseId,
        'message': message,
        'dueAtUtc': dueAt?.toUtc().toIso8601String(),
      },
    );
  }
}
